torch.ops.aten.*: sort aten ops before jit overloads

davidberard98 opened the pull request:
Summary:
In fbcode, aten and jit ops can get registered in different orders depending on build mode. In dev mode, aten is registered first; in opt mode, jit is registered first.

This causes problems in torch.ops.aten.* calls; these calls use `torch._C._jit_get_operation`, which selects an overload based on the inputs to the call. It searches through the overloads for the op with the given name, and chooses the first one that matches the input types. "First" depends on whether aten or jit ops were registered first - e.g. in `test_both_scalars_cuda` in opt mode, it chooses `add.complex` and returns a complex value.

We also saw this issue in https://github.com/pytorch/pytorch/pull/103576.

This PR sorts the list of overloads first, putting the aten ops first.

Differential Revision: D48304930

ezyang reviewed (approved):
Test?

davidberard98 commented:
@ezyang added a test that would have caught this in fbcode, but IDK how to repro the different registration order issue - so it will only be a useful test in the cases where the registration order happens to put JIT overloads ahead of aten overloads

eellison commented on torch/csrc/jit/runtime/operator.h:
```diff
@@ -249,6 +249,8 @@ TORCH_API std::string canonicalSchemaString(const FunctionSchema& schema);
 TORCH_API const std::vector<std::shared_ptr<Operator>> getAllOperators();
 TORCH_API const std::vector<std::shared_ptr<Operator>>& getAllOperatorsFor(
     Symbol name);
+TORCH_API std::vector<std::shared_ptr<Operator>> getAllOperatorsForAtenFirst(
```
nit: if this only called in python files, and only has one use, it's might be better to define it there instead of increasing libtorch's build size.

eellison reviewed (approved):
looks good

davidberard98 commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -1590,7 +1590,7 @@ void initJITBindings(PyObject* module) {
         try {
           auto symbol = Symbol::fromQualString(op_name);
-          auto operations = getAllOperatorsFor(symbol);
+          auto operations = getAllOperatorsForAtenFirst(symbol);
```
good point, moved it

davidberard98 committed:
torch.ops.aten.*: sort aten ops before jit overloads (#107138)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107138

In fbcode, aten and jit ops can get registered in different orders depending on build mode. In dev mode, aten is registered first; in opt mode, jit is registered first.

This causes problems in torch.ops.aten.* calls; these calls use `torch._C._jit_get_operation`, which selects an overload based on the inputs to the call. It searches through the overloads for the op with the given name, and chooses the first one that matches the input types. "First" depends on whether aten or jit ops were registered first - e.g. in `test_both_scalars_cuda` in opt mode, it chooses `add.complex` and returns a complex value.

We also saw this issue in https://github.com/pytorch/pytorch/pull/103576.

This PR sorts the list of overloads first, putting the aten ops first.

Reviewed By: ezyang, eellison

Differential Revision: D48304930

fbshipit-source-id: 40b19d50543a5e5101c57d686dd75b03a8d39d42
